Shopify/ruby-lsp
# Add go to definition for classes and modules

### Motivation

Closes #199

Closing the related issue since the exploration is complete. This PR adds go to definition for classes and modules, which is currently what's covered in the index.

### Implementation

The idea is to find constant/constant path nodes and look for them inside the index and then allow jumping to those.

If the project uses Sorbet, we avoid returning locations for classes and modules that are defined inside the project - since Sorbet can find those even for `typed: false` files. But we still return all results for definitions from gems, which should provide a nice merged experience.

### Automated Tests

Added a test for this.

### Manual Tests

1. Boot the LSP on this branch
2. Try to go to the definition of a class defined in the Ruby LSP itself (e.g.: `EventEmitter`)
3. Verify you jump directly to it, without having multiple options (this is coming from Sorbet only)
4. Try to go to definition of a class defined in a gem (e.g.: SyntaxTree::Const)
5. Verify you get both RBIs and the actual gem source as results
